lib/system/database.rb

```diff
@@ -151,17 +151,9 @@ def adapter_module
 if System::Database.oracle? && defined?(ActiveRecord::ConnectionAdapters::OracleEnhancedAdapter::DatabaseTasks)
   ActiveRecord::ConnectionAdapters::OracleEnhancedAdapter::DatabaseTasks.class_eval do
     prepend(Module.new do
-
+      # Overwrites OracleEnhancedAdapter's create method to prevent Database manipulation using the SYSTEM user.
       def create
-        super
-        connection.execute "GRANT create trigger TO #{username}"
-        connection.execute "GRANT create procedure TO #{username}"
-      end
-
-      protected
-
-      def username
-        @config['username']
```

Comment:
I think we should let SYSTEM user and password be for people willing to supply them. Especially locally I'd like to be able to reset database simply with `db:reset`. Also the `Makefile` will be unable to setup database properly without SYSTEM user's password.

So my request is to add the necessary changes and docs to run without SYSTEM user but also allow us to run with it for the benefit of not performing manual commands.

Reply:
Interesting. This could also work as a failsafe for these changes.
IF `ORACLE_SYSTEM_PASSWORD` is provided, we don't overwrite OracleEnhancedAdapter's create method, which will create/alter the non-SYSTEM user and grant permissions. Now, if it's not provided, we just overwrite the method.
I'll add the change and also add a comment saying that it should be used for development purposes.

Reply:
What is the reason to override the method? If password was not provided, it will not change the database anyway. So why override it?

Reply:
This is the method we're overwriting: [ActiveRecord::ConnectionAdapters::OracleEnhancedAdapter::DatabaseTasks#create](https://github.com/rsim/oracle-enhanced/blob/4ac11553d1e31297b1f7ad421883c9e398e5e9ff/lib/active_record/connection_adapters/oracle_enhanced/database_tasks.rb#L15-L34). If the `ORACLE_SYSTEM_PASSWORD` ENV variable is not provided, it will fail when `rails db:create` is called.

Reply:
We already rid the `create` task in `Rakefile`. It seems a little cleaner to get rid of `db:create` when needed instead of messing with adapter internal details. That may possibly break if newer versions refactored their internal API.

Reply:
I don't get it. What do you mean by "We already rid the create task in Rakefile"?
And are you suggesting to remove this overwrite altogether? If we do so, we won't remove the need for `ORACLE_SYSTEM_PASSWORD`. Also, currently, we overwrite it to add a couple extra GRANTs.

Reply:
Oh I see this is our code. What I meant is that we clear actually the `db:drop` task when needed in `Rakefile`. So I meant to leave `#create` alone and empty `db:create` rake task instead when running on oracle. My understanding is that this `#create` method above is only used when creating a database. If we don't create a database then we will never call this `#create` method. 

But given this is our override already, then maybe it doesn't matter much and might even be better self-contained if you do it here. I just don't understand, why do we need the line `establish_connection(@config)`?

[1] https://github.com/3scale/porta/blob/6827dfa6a0343c16224e232026ea4a6c1d68f20e/Rakefile#L82

Reply:
> why do we need the line establish_connection(@config)?

That's there as the minimum amount of code to be run during a `db:create` task. It just tries to connect to the database and raises an error if it fails. So it's only the "bad path" of a `db:create`.
My idea was to keep failing fast if something is wrong.

Reply:
I think that at the point where we try db:create, next things we would do are also db connection related so it will fail pretty fast. IMO will be more clear about our intention to just make method empty.

Reply:
I can change it, of course. But about making it more clear, I think the current approach is clearer because [1] we don't leave an empty method (of course there'll be a comment) and [2] we keep a consistent behaviour between Adapters when running `db:create`, as it can be executed by itself.

Reply:
IMO making a connection is needless but not a hard opinion so not blocking. 

Actually it makes sense for the method to fail if database does not exist. Then it will be consistent that if we end-up without a database, then it will show it.

I'd only request to actually output a warning that we are not actually creating a database, but using the existing one because SYSTEM_PASSWORD was not provided. In this way things should be more discoverable for the unsuspecting.

Reply:
> I'd only request to actually output a warning that we are not actually creating a database, but using the existing one because SYSTEM_PASSWORD was not provided. In this way things should be more discoverable for the unsuspecting.

Sounds great! Let me work on that.

Reply:
@akostadinov updated in 2053e38.